src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -434,32 +448,79 @@ GraphPatternOperation Visitor::visit(Parser::OptionalGraphPatternContext* ctx) {
   return GraphPatternOperation{parsedQuery::Optional{std::move(pattern)}};
 }
 
-// ____________________________________________________________________________________
+// Parsing for the `serviceGraphPattern` rule.
+parsedQuery::Service Visitor::visit(Parser::ServiceGraphPatternContext* ctx) {
+  // If SILENT is specified, report that we do not support it yet.
+  //
+  // TODO: Support it, it's not hard. The semantics of SILENT is that if no
+  // result can be obtained from the remote endpoint, then do as if the SERVICE
+  // clause would not be there = the result is the neutral element.
+  if (ctx->SILENT()) {
+    reportNotSupported(ctx, "SILENT modifier in SERVICE is");
+  }
+  // Get the IRI and if a variable is specified, report that we do not support
+  // it yet.
+  //
+  // NOTE: According to the grammar, this should either be a `Variable` or an
+  // `Iri`, but due to (not very good) technical reasons, the `visit` returns a
+  // `std::variant<Variable, GraphTerm>`, where `GraphTerm` is a
+  // `std::variant<Literal, BlankNode, Iri>`, hence the `AD_CONTRACT_CHECK`.
+  //
+  // TODO: Also support variables. The semantics is to make a connection for
+  // each IRI matching the variable and take the union of the results.
+  VarOrTerm varOrIri = visit(ctx->varOrIri());
+  if (std::holds_alternative<Variable>(varOrIri)) {
+    reportNotSupported(ctx->varOrIri(), "Variable endpoint in SERVICE is");
+  }
+  AD_CONTRACT_CHECK(std::holds_alternative<Iri>(std::get<GraphTerm>(varOrIri)));
+  Iri serviceIri = std::get<Iri>(std::get<GraphTerm>(varOrIri));
+  // Parse the body of the SERVICE query. Add the visible variables from the
+  // SERVICE clause to the visible variables so far, but also remember them
+  // separately (with duplicates removed) because we need them in `Service.cpp`
+  // when computing the result for this operation.
+  std::vector<Variable> visibleVariablesSoFar = std::move(visibleVariables_);
+  parsedQuery::GraphPattern graphPattern = visit(ctx->groupGraphPattern());
+  std::vector<Variable> visibleVariablesServiceQuery =
+      ad_utility::removeDuplicates(std::move(visibleVariables_));
+  visibleVariables_ = std::move(visibleVariablesSoFar);
```

Comment:
This block very similarly appears in the parsing of subqueries, 
I think we can factor it out
(visit without changing the visible variables, and returning the visible variables separately).  Then only the `insert` remains (which we don't have for the subqueries).